### PR TITLE
increase the default timeout for PTL tests

### PR DESCRIPTION
**Pull request by dinesh1306**
#### Bug/feature Description
* Increase the default timeout for PTL tests to 600sec.
* [PP-1047](https://pbspro.atlassian.net/browse/PP-1047)*

#### Affected Platform(s)
* All

#### Cause / Analysis / Design
* sometime if setup takes longer time in some tests where there are too many jobs or nodes to cleanup then it affects test execution time. default time for each test case execution is 180sec including setUp and tearDown. If setUp and test execution is not completed in 180sec then test got timeout.

#### Solution Description
* Increase the default timeout for PTL tests to 600sec. if timeout decorator have value less than 600sec(Default testcase timeout) then PTL consider testcase timeout value as Default testcase timeout otherwise consider value provided by decorator. 

#### Testing logs/output
* [res_TestDefaultTimeout.txt](https://github.com/PBSPro/pbspro/files/2648964/res_TestDefaultTimeout.txt)

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

**Comment by bhroam** [comment]
Is this the right answer, or is just not counting the PbsTestSuite.setUp() and PbsTestSuite.tearDown() time?  I would include the part of setUp() and tearDown() that the test writer wrote themselves.

File: test/tests/selftest/pbs_default_timeout.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        """
+        Verify that test not timedout after 180 sec.
+        i.e. after previous default timeout value
+        """
+        time.sleep(300)
+
+    @timeout(200)
+    def test_timeout_decorator_less_default_value(self):
+        """
+        If timeout decorator value is less than 600 then testcase timeout value
```
Comment (anamikau) [inline test/tests/selftest/pbs_default_timeout.py]: reword it to say "If timeout decorator value is less than 600 then default test case timeout is considered"

**Comment by bhroam** [comment]
@anamikau All of setUp() and tearDown() are included.  I'm suggesting we don't include the time for PbsTestSuite.setUp() and PbsTestSuite.tearDown().  The example given is if setUp() has to delete a whole bunch of jobs.  That happens in PbsTestSuite.setUp().  If time spent in that function wasn't included in the timeout, the 180s might be fine.

**Comment by anamikau** [comment]
@agrawalravi90 and @bhroam you are absolutely right. This is a temporary fix to avoid failures that we are observing during the internal testing. For long term we would want to remove setUp and tearDown from the timeout.
I suggest we do not increase the scope of this one and we can resolve the actual problem in different PR.

**Comment by agrawalravi90** [comment]
> For long term we would want to remove setUp and tearDown from the timeout.

Actually that's not what we had discussed. We had just decided to move the reverting to defaults action to tearDown, that way the timeout will correctly apply to the test case which submits the large number of jobs and we don't have to change the default timeout. It's important for the timeout to apply to setUp and tearDown as well because those functions can also get stuck and we'd wanna timeout and bail if that happens.

Is this very urgent to fix? Could we just not run the tests which cause timeout issues until we fix this correctly? It shouldn't be that big of a change to do this properly, I fear that a duct tape fix now will remove the motivation to fix this the right way.

**Comment by bhroam** [comment]
@agrawalravi90 I'm not sure you want to move revert_to_defaults() to tearDown().  Right now it is in setUp().  If setUp() times out, the system is in some in between state.  The next call to setUp() will call revert_to_defaults() and continue the process.  If you move it to tearDown(), it will time out and leave the complex in the in between state which could cause the next test to fail.  I guess it is bad in either case.  Today we will timeout the following test.  In the proposed changes, we'll timeout the current test and fail the following test (or worse, not find a bug when there is one).

**Comment by agrawalravi90** [comment]
@bhroam, actually i didn't lay out the whole design that was discussed, we had discussed putting a "conditional" revert_to_default in setUp() which will be executed only if the previous tearDown() failed. This is why I had proposed that we move this to a forum discussion.

**Comment by agrawalravi90** [comment]
The nice thing about moving revert to tearDown is that the test that is creating a lot of cruft to be cleaned up later can create a custom timeout for itself and ensure that it doesn't time out.

**Comment by anamikau** [comment]
@dinesh1306 please fix codacy error.

@bhroam and @agrawalravi90, can we just have the short term fix in this ticket and take the longer term approach in a different PR?

**Comment by anamikau** [comment]
> > For long term we would want to remove setUp and tearDown from the timeout.
> 
> Actually that's not what we had discussed. We had just decided to move the reverting to defaults action to tearDown, that way the timeout will correctly apply to the test case which submits the large number of jobs and we don't have to change the default timeout. It's important for the timeout to apply to setUp and tearDown as well because those functions can also get stuck and we'd wanna timeout and bail if that happens.

Sorry you are right. I had to look up the notes again to recall.
> 
> Is this very urgent to fix? Could we just not run the tests which cause timeout issues until we fix this correctly? It shouldn't be that big of a change to do this properly, I fear that a duct tape fix now will remove the motivation to fix this the right way.

Skipping these tests would be tricky as these failures are very intermittent on different systems. Again this fix is not going to solve the problem 100% so we need to work on the long term approach sooner.

**Comment by bhroam** [comment]
@anamikau if the short term fix is really important, then yeah I can see it.  I can see Ravi's point though.  After the short term fix, would the long term fix be not as high priority?

**Comment by anamikau** [comment]
Thanks @bhroam. This fix should not change the priority for the other bug.

File: test/fw/ptl/utils/pbs_testsuite.py [hunk 2]
```diff
@@ -155,6 +156,9 @@ def timeout(val):
     """
     Decorator to set timeout value of test case
     """
+    if val < DEFAULT_TESTCASE_TIMEOUT:
+        val = DEFAULT_TESTCASE_TIMEOUT
```
Comment (agrawalravi90) [inline test/fw/ptl/utils/pbs_testsuite.py]: why are you doing this? If a test case explicitly says it needs x seconds of timeout, why do you wanna change it to be a minimum of DEFAULT_TESTCASE_TIMEOUT ? I think we should leave custom timeouts alone.

File: test/tests/selftest/pbs_default_timeout.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    """
+    Test suite to verify increase default testcase timeout working properly.
+    """
+
+    def test_default_timeout(self):
+        """
+        Verify that test not timedout after 180 sec.
+        i.e. after previous default timeout value
+        """
+        time.sleep(300)
```
Comment (agrawalravi90) [inline test/tests/selftest/pbs_default_timeout.py]: How about this: Catch the TimeOut exception and mark the test as 'failed' if the exception is caught in 300 seconds? Right now the test is not really very helpful, it will itself timeout while testing timeout, it should either pass or fail correctly.

File: test/tests/selftest/pbs_default_timeout.py (hunk 1 above)
Comment (agrawalravi90) [reply test/tests/selftest/pbs_default_timeout.py]: I strongly feel that PTL should not modify custom timeouts provided by tests internally.

File: test/tests/selftest/pbs_default_timeout.py (hunk 3 above)
Comment (agrawalravi90) [reply test/tests/selftest/pbs_default_timeout.py]: And instead of 300 seconds, you should probably use something like DEFAULT_TESTCASE_TIMEOUT / 2

**Comment by agrawalravi90** [comment]
@anamikau , alright sounds good, I'm ok putting in this fix if it is urgent.

File: test/fw/ptl/utils/pbs_testsuite.py (hunk 2 above)
Comment (dinesh1306) [reply test/fw/ptl/utils/pbs_testsuite.py]: @agrawalravi90 , previously timeout value was 180 and there are some tests have value less than 600 and greater than 180 in timeout tag, to make affect on those tests also i added above condition.

**Comment by agrawalravi90** [comment]
Ok, let's call it 'minimum_testcase_timeout' instead of 'default_testcase_timeout' then. Yes, a warning message would be good.

**Comment by anamikau** [comment]
Thanks @agrawalravi90. your suggestion looks fine to me. 
@dinesh1306 please address Ravi's comments above and also fixed codacy errors.

**Comment by dinesh1306** [comment]
@anamikau @agrawalravi90 , timeout decorator evaluate initially in test suite. If we put warning message in timeout decorator function then it will comes initially in testsuite. Please let me know can i add warning message or not.

**Comment by dinesh1306** [comment]
@anamikau @agrawalravi90 , I have incorporate review comments other than warning message. Please review.

File: test/tests/selftest/pbs_default_timeout.py [hunk 4]
```diff
@@ -36,32 +36,57 @@
...
+from ptl.utils.plugins.ptl_test_runner import TimeOut
 
 
 class TestDefaultTimeout(TestSelf):
     """
     Test suite to verify increase default testcase timeout working properly.
     """
 
+    @staticmethod
+    def common_sleep_function(sleep_val):
```
Comment (agrawalravi90) [inline test/tests/selftest/pbs_default_timeout.py]: why did you need to create this function? couldn't you have just called time.sleep() ?

File: test/tests/selftest/pbs_default_timeout.py [hunk 5]
```diff
@@ -36,32 +36,57 @@
...
         """
         Verify that test not timedout after 180 sec.
         i.e. after previous default timeout value
         """
-        time.sleep(300)
+        try:
+            self.common_sleep_function(MINIMUM_TESTCASE_TIMEOUT/2)
+        except TimeOut:
+            mssg = "Test Timeout before complete default-testcase-timeout"
+            raise TimeOut(mssg)
```
Comment (agrawalravi90) [inline test/tests/selftest/pbs_default_timeout.py]: don't raise TimeOut exception, that would have been the same thing as letting it timeout without catching it. we want to just let the user know that the test failed. So, maybe do "self.fail(\<error message>)" instead of raising TimeOut?

File: test/tests/selftest/pbs_default_timeout.py [hunk 6]
```diff
@@ -36,32 +36,57 @@
...
+
+    def test_timeout_greater_default_value(self):
+        """
+        Verify that test timedout after 600 sec.
+        """
+        mssg = 'Timed out after 600 second'
```
Comment (agrawalravi90) [inline test/tests/selftest/pbs_default_timeout.py]: shouldn't it be 600 seconds (not second) ?

File: test/tests/selftest/pbs_default_timeout.py [hunk 7]
```diff
@@ -36,32 +36,57 @@
...
-        If timeout decorator value is less than 600 then testcase timeout value
-        consider as default timeout value
+        If timeout decorator value is less than 600 then
+        default testcase timeout is considered
         """
-        time.sleep(300)
+        try:
+            self.common_sleep_function(MINIMUM_TESTCASE_TIMEOUT/2)
+        except TimeOut:
+            raise TimeOut('Test Timeout before complete expected time')
```
Comment (agrawalravi90) [inline test/tests/selftest/pbs_default_timeout.py]: same comment here, don't raise TimeOut

File: test/tests/selftest/pbs_default_timeout.py [hunk 8]
```diff
@@ -36,32 +36,57 @@
...
     def test_timeout_decorator_greater_default_value(self):
         """
         If timeout decorator value is greater than 600 then
         testcase timeout value consider as timeout decorator value
         """
-        time.sleep(700)
+        try:
+            self.common_sleep_function(MINIMUM_TESTCASE_TIMEOUT+1)
+        except TimeOut:
+            raise TimeOut('Test Timeout before complete expected time')
```
Comment (agrawalravi90) [inline test/tests/selftest/pbs_default_timeout.py]: same comment here, don't raise TimeOut

File: test/fw/ptl/utils/pbs_testsuite.py [hunk 9]
```diff
@@ -156,8 +156,8 @@ def timeout(val):
     """
     Decorator to set timeout value of test case
     """
-    if val < DEFAULT_TESTCASE_TIMEOUT:
-        val = DEFAULT_TESTCASE_TIMEOUT
+    if val < MINIMUM_TESTCASE_TIMEOUT:
+        val = MINIMUM_TESTCASE_TIMEOUT
```
Comment (agrawalravi90) [inline test/fw/ptl/utils/pbs_testsuite.py]: this is where you might wanna add a warning message

File: test/tests/selftest/pbs_default_timeout.py (hunk 4 above)
Comment (dinesh1306) [reply test/tests/selftest/pbs_default_timeout.py]: @agrawalravi90 , Initially i did the same but codacy giving error "Method could be a function". to incorporate it i make static common function and called in every test.

File: test/fw/ptl/utils/pbs_testsuite.py (hunk 9 above)
Comment (dinesh1306) [reply test/fw/ptl/utils/pbs_testsuite.py]: I will add warning message here, but timeout decorator evaluate initially in test suite it will comes initially in testsuite not at the test level.

File: test/tests/selftest/pbs_default_timeout.py (hunk 6 above)
Comment (dinesh1306) [reply test/tests/selftest/pbs_default_timeout.py]: This message is from timeout_handler function.

def timeout_handler(signum, frame):
    raise TimeOut('Timed out after %s second' % timeout)

File: test/fw/ptl/utils/pbs_testsuite.py (hunk 9 above)
Comment (anamikau) [reply test/fw/ptl/utils/pbs_testsuite.py]: @dinesh1306 thanks for the info. I suggest still adding a generic message mentioning that some of the tests are updated to use minimum_test_timeout

File: test/tests/selftest/pbs_default_timeout.py (hunk 6 above)
Comment (agrawalravi90) [reply test/tests/selftest/pbs_default_timeout.py]: i think it's safe to change the string to 'seconds', changes of the timeout period being 1 second is pretty small

File: test/tests/selftest/pbs_default_timeout.py (hunk 5 above)
Comment (bhroam) [reply test/tests/selftest/pbs_default_timeout.py]: Strange, because that is not what the instructions say: https://docs.python.org/2/library/unittest.html#skipping-tests-and-expected-failures
It says that it's more like a skip since it's an expected failure.

File: test/tests/selftest/pbs_default_timeout.py (hunk 4 above)
Comment (agrawalravi90) [reply test/tests/selftest/pbs_default_timeout.py]: no, i meant to say that you don't need this function altogether, you can just call time.sleep() instead of calling this function in your code

File: test/tests/selftest/pbs_default_timeout.py (hunk 5 above)
Comment (agrawalravi90) [reply test/tests/selftest/pbs_default_timeout.py]: If you mean this line: " In addition, it supports marking a test as an “expected failure,” a test that is broken and will fail, but shouldn’t be counted as a failure on a TestResult."

Then that refers to the decorator "@unittest.expectedFailure", not the unittest.fail() function. If you look at the logs that I posted, self.fail() basically raises an AssertionError

Comment (bhroam) [reply test/tests/selftest/pbs_default_timeout.py]: I read it a bit too fast.  Thanks for pointing it out.  We can stop using self.assertTrue(False) now.  This will work for us.  Thanks for testing it out @agrawalravi90

**Comment by anamikau** [comment]
Marking as WIP as @dinesh1306 is on leave and there is no one who can work on this.

File: test/tests/selftest/pbs_default_timeout.py (hunk 6 above)
Comment (agrawalravi90) [reply test/tests/selftest/pbs_default_timeout.py]: I'm okay with you leaving this as is as the message is generated by PTL, other tests might also be searching for the same string, so we can deal with that separately.

**Comment by dinesh1306** [comment]
@agrawalravi90 , @anamikau I have incorporate review comments. Please Take a look.

**Comment by dinesh1306** [comment]
[result.txt](https://github.com/PBSPro/pbspro/files/2726096/result.txt)
Latest Execution Result

File: test/fw/ptl/utils/pbs_testsuite.py (hunk 9 above)
Comment (dinesh1306) [reply test/fw/ptl/utils/pbs_testsuite.py]: @anamikau , I have added message.

File: test/tests/selftest/pbs_default_timeout.py [hunk 10]
```diff
@@ -87,6 +84,7 @@ def test_timeout_decorator_greater_default_value(self):
         testcase timeout value consider as timeout decorator value
         """
         try:
-            self.common_sleep_function(MINIMUM_TESTCASE_TIMEOUT+1)
+            time.sleep(MINIMUM_TESTCASE_TIMEOUT+1)
         except TimeOut:
-            raise TimeOut('Test Timeout before complete expected time')
+            mssg = 'Test Timeout before complete expected time'
```
Comment (agrawalravi90) [inline test/tests/selftest/pbs_default_timeout.py]: actually, there can be a race condition here: if the setUp takes a long time (e.g - if it has to delete a large number of jobs from the previous test), then the time taken by setUp + time.sleep(MINIMUM_TESTCASE_TIMEOUT+1) might actually cross timeout(800). So, instead of failing the test outright, have a assert to check that the timeout was for 800, if not then fail the test.

File: test/tests/selftest/pbs_default_timeout.py [hunk 11]
```diff
@@ -44,28 +44,24 @@ class TestDefaultTimeout(TestSelf):
...
         Verify that test not timedout after 180 sec.
         i.e. after previous default timeout value
         """
         try:
-            self.common_sleep_function(MINIMUM_TESTCASE_TIMEOUT/2)
+            time.sleep(MINIMUM_TESTCASE_TIMEOUT/2)
         except TimeOut:
-            mssg = "Test Timeout before complete default-testcase-timeout"
-            raise TimeOut(mssg)
+            mssg = 'Test Timeout before complete default-testcase-timeout'
```
[the remark this answers is not in the document]
Comment (agrawalravi90) [reply test/tests/selftest/pbs_default_timeout.py]: same comment here about the race condition, please check my comment and add a check here to see if the timeout was for the correct duration or not.

File: test/tests/selftest/pbs_default_timeout.py [hunk 12]
```diff
@@ -76,9 +72,10 @@ def test_timeout_decorator_less_default_value(self):
         default testcase timeout is considered
         """
         try:
-            self.common_sleep_function(MINIMUM_TESTCASE_TIMEOUT/2)
+            time.sleep(MINIMUM_TESTCASE_TIMEOUT/2)
         except TimeOut:
-            raise TimeOut('Test Timeout before complete expected time')
+            mssg = 'Test Timeout before complete expected time'
```
[the remark this answers is not in the document]
Comment (agrawalravi90) [reply test/tests/selftest/pbs_default_timeout.py]: same comment here about the race condition

File: test/tests/selftest/pbs_default_timeout.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+        try:
+            mssg = 'sleeping for %ssec and minimum-testcase-timeout is %ssec' \
+                   % (MINIMUM_TESTCASE_TIMEOUT+1, MINIMUM_TESTCASE_TIMEOUT)
+            self.logger.info(mssg)
+            time.sleep(MINIMUM_TESTCASE_TIMEOUT+1)
+        except TimeOut, e:
+            mssg = 'Timed out after 600 second'
+            self.assertEqual(mssg, e.message)
+        else:
+            self.fail('Test is not timeout after minimum timeout period')
```
Comment (agrawalravi90) [inline test/tests/selftest/pbs_default_timeout.py]: How about: "Test did not timeout after the minimum timeout period of %d"%MINIMUM_TESTCASE_TIMEOUT

File: test/tests/selftest/pbs_default_timeout.py [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+        default testcase timeout is considered
+        """
+        try:
+            mssg = 'sleeping for %ssec and minimum-testcase-timeout is %ssec' \
+                   % (MINIMUM_TESTCASE_TIMEOUT/2, MINIMUM_TESTCASE_TIMEOUT)
+            self.logger.info(mssg)
+            time.sleep(MINIMUM_TESTCASE_TIMEOUT/2)
+        except TimeOut, e:
+            mssg = 'Timed out after 600 second'
+            self.assertEqual(mssg, e.message)
```
Comment (agrawalravi90) [inline test/tests/selftest/pbs_default_timeout.py]: Please add an error message to assert and do this for the rest of the code as well.

**Comment by dinesh1306** [comment]
@agrawalravi90 , I have incorporate review comments. Please take a look.

File: test/tests/selftest/pbs_default_timeout.py [hunk 15]
```diff
@@ -55,8 +55,9 @@ def test_default_timeout(self):
             self.logger.info(mssg)
             time.sleep(MINIMUM_TESTCASE_TIMEOUT/2)
         except TimeOut, e:
-            mssg = 'Timed out after 600 second'
-            self.assertEqual(mssg, e.message)
+            mssg = 'Timed out after %s second' % MINIMUM_TESTCASE_TIMEOUT
+            err_mssg = 'failed due to mismatch in expected message'
```
Comment (agrawalravi90) [inline test/tests/selftest/pbs_default_timeout.py]: this log message doesn't really explain the main reason why the test failed. It might be more useful to say something like "the test timed out for an incorrect timeout duration". If you agree, then please make change the log message in the rest of the code as well.

**Comment by dinesh1306** [comment]
@agrawalravi90 ,  i made changes according to your comment. Please review.

File: test/tests/selftest/pbs_default_timeout.py [hunk 16]
```diff
@@ -50,20 +50,28 @@ def test_default_timeout(self):
...
         """
         try:
+            mssg = 'sleeping for %ssec and minimum-testcase-timeout is %ssec' \
+                   % (MINIMUM_TESTCASE_TIMEOUT/2, MINIMUM_TESTCASE_TIMEOUT)
+            self.logger.info(mssg)
             time.sleep(MINIMUM_TESTCASE_TIMEOUT/2)
-        except TimeOut:
-            mssg = 'Test Timeout before complete default-testcase-timeout'
-            self.fail(mssg)
+        except TimeOut, e:
```
Comment (bhroam) [inline test/tests/selftest/pbs_default_timeout.py]: Please use the modern style of except: except TimeOut as e:

File: test/tests/selftest/pbs_default_timeout.py [hunk 17]
```diff
@@ -72,10 +80,13 @@ def test_timeout_decorator_less_default_value(self):
...
         """
         try:
+            mssg = 'sleeping for %ssec and minimum-testcase-timeout is %ssec' \
+                   % (MINIMUM_TESTCASE_TIMEOUT/2, MINIMUM_TESTCASE_TIMEOUT)
+            self.logger.info(mssg)
             time.sleep(MINIMUM_TESTCASE_TIMEOUT/2)
-        except TimeOut:
-            mssg = 'Test Timeout before complete expected time'
-            self.fail(mssg)
+        except TimeOut, e:
```
Comment (bhroam) [inline test/tests/selftest/pbs_default_timeout.py]: same here, use 'as'

File: test/tests/selftest/pbs_default_timeout.py [hunk 18]
```diff
@@ -84,7 +95,10 @@ def test_timeout_decorator_greater_default_value(self):
...
         """
         try:
+            mssg = 'sleeping for %ssec and minimum-testcase-timeout is %ssec' \
+                   % (MINIMUM_TESTCASE_TIMEOUT+1, 800)
+            self.logger.info(mssg)
             time.sleep(MINIMUM_TESTCASE_TIMEOUT+1)
-        except TimeOut:
-            mssg = 'Test Timeout before complete expected time'
-            self.fail(mssg)
+        except TimeOut, e:
```
Comment (bhroam) [inline test/tests/selftest/pbs_default_timeout.py]: here as well, use 'as'

**Comment by bhroam** [comment]
@dinesh1306 thanks for fixing my suggestions.  I didn't do a full review, so please continue without my signoff.

**Comment by dinesh1306** [comment]
@agrawalravi90 , @anamikau Thanks for the review. Branch is rebased and squashed. Ready for merge.
[pbs_final_result.txt](https://github.com/PBSPro/pbspro/files/2749594/pbs_final_result.txt)
